sql/core/src/test/scala/org/apache/spark/sql/sources/FSBasedRelationSuite.scala

```diff
@@ -0,0 +1,425 @@
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.sources
+
+import java.io.IOException
+
+import scala.collection.mutable.ArrayBuffer
+
+import org.apache.hadoop.fs.{FileSystem, Path}
+
+import org.apache.spark.rdd.RDD
+import org.apache.spark.sql.types._
+import org.apache.spark.sql.{Row, SQLContext, SaveMode}
+import org.apache.spark.util.Utils
+
+class SimpleFSBasedSource extends RelationProvider {
+  override def createRelation(
+      sqlContext: SQLContext,
+      parameters: Map[String, String]): BaseRelation = {
+    SimpleFSBasedRelation(parameters)(sqlContext)
+  }
+}
+
+case class SimpleFSBasedRelation
+    (parameter: Map[String, String])
+    (val sqlContext: SQLContext)
+  extends FSBasedRelation {
+
+  class SimpleOutputWriter extends OutputWriter {
+    override def write(row: Row): Unit = TestResult.writtenRows += row
+  }
+
+  override val path = parameter("path")
+
+  override def dataSchema: StructType =
+    DataType.fromJson(parameter("schema")).asInstanceOf[StructType]
+
+  override def buildScan(
+      requiredColumns: Array[String],
+      filters: Array[Filter],
+      inputPaths: Array[String]): RDD[Row] = {
+    val basePath = new Path(path)
+    val fs = basePath.getFileSystem(sqlContext.sparkContext.hadoopConfiguration)
+
+    // Used to test queries like "INSERT OVERWRITE tbl SELECT * FROM tbl". Source directory
+    // shouldn't be removed before scanning the data.
+    assert(inputPaths.map(new Path(_)).forall(fs.exists))
+
+    TestResult.requiredColumns = requiredColumns
+    TestResult.filters = filters
+    TestResult.inputPaths = inputPaths
+
+    Option(TestResult.rowsToRead).getOrElse(sqlContext.emptyResult)
+  }
+
+  override def outputWriterClass: Class[_ <: OutputWriter] = classOf[SimpleOutputWriter]
+}
+
+object TestResult {
+  var requiredColumns: Array[String] = _
+  var filters: Array[Filter] = _
+  var inputPaths: Array[String] = _
+  var rowsToRead: RDD[Row] = _
+  var writtenRows: ArrayBuffer[Row] = ArrayBuffer.empty[Row]
+
+  def reset(): Unit = {
+    requiredColumns = null
+    filters = null
+    inputPaths = null
+    rowsToRead = null
+    writtenRows.clear()
+  }
+}
+
+class FSBasedRelationSuite extends DataSourceTest {
```

Comment:
Please also add error message tests for when the functions are called on a datasource that does not support partitioning.